src/utilities/geometry/Polyhedron.cpp

```diff
@@ -58,136 +67,246 @@ bool Surface3dEdge::operator!=(const Surface3dEdge& other) const {
+      if (m_polyhedronVolume < 0) {
+        LOG(Error, "It seems that ALL surfaces are reversed");
+        m_hasAnySurfaceWithIncorrectOrientation = true;
+        m_isCompletelyInsideOut = true;
       }
     }
   }
-  return uniqVertices;
+};
+
+bool Polyhedron::hasEdgesNot2() const {
+  return std::any_of(m_surfaces.cbegin(), m_surfaces.cend(), [](const auto& surface) {
+    return std::any_of(surface.edges.cbegin(), surface.edges.cend(), [](const auto& edge) { return edge.count() != 2; });
+  });
 }
 
-std::vector<Surface3dEdge> Polyhedron::edgesNotTwoForEnclosedVolumeTest(const Polyhedron& volumePoly) {
+bool Polyhedron::hasAnySurfaceWithIncorrectOrientation() const {
+  return m_hasAnySurfaceWithIncorrectOrientation;
+}
 
-  std::vector<Surface3dEdge> uniqueSurface3dEdges;
-  uniqueSurface3dEdges.reserve(volumePoly.numVertices());
+bool Polyhedron::hasAddedColinearPoints() const {
+  return m_hasAddedColinearPoints;
+}
 
-  // construct list of unique edges
-  int surfNum = 0;
-  for (const auto& surface : volumePoly.m_surfaces) {
-    LOG(Debug, "Surface: " << surface.name);
-    const auto& vertices = surface.vertices;
-    for (auto it = vertices.begin(); it != vertices.end(); ++it) {
-      auto itnext = std::next(it);
-      if (itnext == std::end(vertices)) {
-        itnext = std::begin(vertices);
+bool Polyhedron::isEnclosedVolume() const {
+  return m_isEnclosedVolume;
+}
+
+bool Polyhedron::isCompletelyInsideOut() const {
+  return m_isCompletelyInsideOut;
+}
+
+void Polyhedron::performEdgeMatching() {
+
+  m_hasAnySurfaceWithIncorrectOrientation = false;
+
+  for (size_t i = 0; i < m_surfaces.size(); ++i) {
+    for (size_t j = 0; j < m_surfaces.size(); ++j) {
+      if (i == j) {
+        continue;
       }
-      Surface3dEdge thisSurface3dEdge(*it, *itnext, surface, surfNum);
-      auto itFound = std::find(uniqueSurface3dEdges.begin(), uniqueSurface3dEdges.end(), thisSurface3dEdge);
-      if (itFound == uniqueSurface3dEdges.end()) {
-        LOG(Debug, "NOT FOUND: " << thisSurface3dEdge);
-        uniqueSurface3dEdges.emplace_back(std::move(thisSurface3dEdge));
-      } else {
-        LOG(Debug, "    FOUND: " << thisSurface3dEdge);
-        itFound->appendSurface(surface);
+      auto& surface1 = m_surfaces[i];
+      auto& surface2 = m_surfaces[j];
+      for (Surface3dEdge& edge1 : surface1.edges) {
+        for (Surface3dEdge& edge2 : surface2.edges) {
+          if (edge1 == edge2) {
+            if (std::find(edge1.allSurfNums().begin(), edge1.allSurfNums().cend(), edge2.firstSurfNum()) == edge1.allSurfNums().end()) {
+              edge1.appendSurface(surface2);
+              edge2.appendSurface(surface1);
+              if (!edge1.reverseEqual(edge2)) {
+                edge1.markConflictedOrientation();
+                edge2.markConflictedOrientation();
+                m_hasAnySurfaceWithIncorrectOrientation = true;
+              }
+            }
+          }
+        }
```

Comment:
Performing edge matching, in one go we pair the edges (later we can check that edge.count() == 2, any Polyhedron which has edges where count() != 2 is not enclosed), and we also check if that edge matches **in reverse order** to the other.

In a polyhedron that has a consistent orientation (typically all faces are in counter clockwise order), each edge must be matched by an edge in the opposite direction.